CONVENTIONS.md

```diff
@@ -0,0 +1,124 @@
+`setColor(newColor)`).
+
+Use "with" for methods that return a copy of an object with a different value for some property. Name the parameter the same
+as the property.
+
+Use "create" for methods that create a new instance of some object. Use "get or create" for methods that create a new
+instance only if one does not already exist. Don't use "get or create" for lazy initialization, though. 
+
+### Serialization
+
+Use "serializer" for objects whose purpose is serializing or deserializing some type of object (`RecipeSerializer`). Use
+"serialize" and "deserialize" for methods only when serializing or deserializing an object other than the one the method is in.
+
+Use "from" for static methods that create an object of the method owner's type (`fromJson`, `fromNbt`, `fromString`). Use "to"
+for methods that convert an object to another type (`toString`, `toLong`, `toNbt`).
+
+Use "read" for non-static methods that load data into a class. Use "write" for methods that save data to an *existing* object
+passed as a parameter.
+
+### Factories and builders
+
+Use "factory" for objects whose purpose is creating other objects.
+
+Use "builder" for objects whose purpose is helping with the creation of an immutable object. Name builder methods the same
+as the field they're setting, without any prefix.
+
+### Collections
+
+Use a plural name for collections and maps rather than the words "list", "set", "array", etc., unless it's a collection of
+collection or there are several collections of different types containing the same objects (`entities`, `entityLists`).
+
+When it's enough, name maps based on the value type. Otherwise, name it in the "`valuesByKeys`" format.
+
+## Mojang names
+
+Use names that match names in strings in the vanilla code, unless that string is outdated or inaccurate. This avoids confusion,
+especially from new modders who may not understand what an class exception message is referring to.
+
+Even if a known Mojang name doesn't appear in any strings, it's a good idea to use it since the official name is a good
+indicator of the the class's actual purpose and makes it less likely the name will have to be changed in the future. For
+example, don't name a class that Mojang calls `BedrockBlock` `NoSpawningBlock`, even if its only purpose is to disable mob
+spawning, because Mojang may decide to override more methods in that class, breaking mods that were using it in an unexpected
+way.
+
+There are however three exceptions to this rule:
+ - Use "world" for what Mojang calls "level" (see https://github.com/FabricMC/yarn/issues/89)
+ - Use "container" for what Mojang calls "menu" (see https://github.com/FabricMC/yarn/issues/386)
+ - Use "inventory" for what Mojang calls "container" (no issue yet, requires renaming "container" first)
+
+## Things to avoid
+
+Don't naming methods based on implementation details. Names should describe what methods do, now how they work.
```

Comment:
Did you mean "Don't name methods based on..."?

Reply:
Yes, fixed.